.github/workflows/checks.yml

```diff
@@ -9,100 +9,111 @@ on:
+  # ...................LINT-FORMAT-TYPECHECK.................. #
+  # ---------------------------------------------------------- #
+
+  lint-format-typecheck:
+    name: Lint - Format - Typecheck
     runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        python-version: [ "3.9" ]
     steps:
+
       - uses: actions/checkout@v3
-        with:
-          submodules: recursive
+
       - name: Install poetry
         run: |
           python -m pip install --upgrade pip
           pip install poetry
           poetry config installer.modern-installation false
-      - name: Set up Python ${{ matrix.python-version }}
+
+      - name: Set up Python 3.9
         uses: actions/setup-python@v4
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: "3.9"
           cache: 'poetry'
+
       - name: Install dependencies
         run: |
           poetry install
 
-  lint:
-    name: Check lock, formatting, linting and typing
-    runs-on: ubuntu-latest
-    needs: setup
-    strategy:
-      matrix:
-        python-version: [ "3.9" ]
-    steps:
-      - uses: actions/checkout@v3
-        with:
-          submodules: recursive
-      # poetry needs to be reinstalled because only commands starting with `poetry` are cached
-      - name: Install poetry
-        run: |
-          python -m pip install --upgrade pip
-          pip install poetry
-          poetry config installer.modern-installation false
-      - name: Set up Python ${{ matrix.python-version }}
-        uses: actions/setup-python@v4
-        with:
-          python-version: ${{ matrix.python-version }}
-          cache: 'poetry'
       - name: Check poetry.lock
         run: |
           poetry lock --check
-      - name: Check formatting with black
-        run: |
-          poetry run poe format_check
-      - name: Analysing the code with pylint
+
+      - name: Lint
         run: |
           poetry run poe lint
-      - name: Typechecking
+
+      - name: Format
+        run: |
+          poetry run poe format_check
+
+      - name: Typecheck
         run: |
           poetry run poe typecheck
 
-  unit-and-e2e-tests:
-    name: Unit and e2e tests
+  # ---------------------------------------------------------- #
+  # ..........................TESTS........................... #
+  # ---------------------------------------------------------- #
+
+  run-tests:
+    name: Tests
     runs-on: ubuntu-latest
-    needs: setup
     strategy:
       matrix:
-        python-version: [ "3.9" ]
+        python-version: [ "3.8", "3.9", "3.10", "3.11" ]
     steps:
+
+      # ====================== SETUP ====================== #
+
       - uses: actions/checkout@v3
+
+      - uses: actions/setup-python@v4
         with:
-          submodules: recursive
+          python-version: "3.9"
+          cache: 'pip'
+
+      - name: Install python 3.9 requirements
+        run: |
+          pip install cairo-lang starknet-devnet
```

Comment:
Probably, we should install cairo-lang and starknet-devnet from pyproject.toml somehow.
Or manually specify versions here (worse option)